### PR TITLE
356 Use tls chain file

### LINKED ISSUE #356: Using LetsEncrypt TLS certificate results in error: "unable to get local issuer certificate"
Build from source Source: commit d745b15768ce0269776db339743f6516b0b5008f using the kanidmd/Dockerfile.

I am trying use a Letsencrypt certificate to secure the idm server.

First issue is that checking the certificate locally with `openssl` already fails.

```
openssl verify -show_chain -CAfile fullchain.pem cert.pem
C = US, O = Let's Encrypt, CN = R3
error 2 at 1 depth lookup: unable to get issuer certificate
error cert.pem: verification failed
```

This can be solved to add Root X3 certificate to the `fullchain.pem`.
Download root x3 and place it in `rootx3.pem`.
```
cat rootx3.pem fullchain.pem > ca.pem
openssl verify -show_chain -CAfile ca.pem cert.pem
cert.pem: OK
Chain:
depth=0: CN = <my FQDN> (untrusted)
depth=1: C = US, O = Let's Encrypt, CN = R3
depth=2: O = Digital Signature Trust Co., CN = DST Root CA X3
```

I copied the created `ca.pem`, `cert.pem` and `privkey.pem` to the container volume.
Also checked with `openssl` in the container that the certificate is valid.

```
podman run --rm -i -t -v kanidmd:/data alpine /bin/sh
/ # apk add openssl
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.13/community/x86_64/APKINDEX.tar.gz
(1/1) Installing openssl (1.1.1i-r0)
Executing busybox-1.32.1-r2.trigger
OK: 6 MiB in 15 packages

/ # cd /data/
/data # openssl verify -show_chain -CAfile ca.pem cert.pem
cert.pem: OK
Chain:
depth=0: CN = <my FQDN> (untrusted)
depth=1: C = US, O = Let's Encrypt, CN = R3
depth=2: O = Digital Signature Trust Co., CN = DST Root CA X3
```

Firefox reports that the connection is secure.
Log shows:
```
[tide::log::middleware] <-- Request received
[tide::log::middleware] Client error --> Response sent
```

When using `./kanidm login -H https://<my FQDN>:8443 -D admin` I get this error.
```
[2021-02-11T15:05:44Z ERROR kanidm_cli::login] Error during authentication init phase: Transport(reqwest::Error { kind: Request, url: Url { scheme: "https", host: Some(Domain("<my FQDN>")), port: Some(8443), path: "/v1/auth", query: None, fragment: None }, source: hyper::Error(Connect, Ssl(Error { code: ErrorCode(1), cause: Some(Ssl(ErrorStack([Error { code: 337047686, library: "SSL routines", function: "tls_process_server_certificate", reason: "certificate verify failed", file: "ssl/statem/statem_clnt.c", line: 1913 }]))) }, X509VerifyResult { code: 20, error: "unable to get local issuer certificate" })) })
```
Log shows: `[tide_rustls::tls_listener] tls error`.

I also tried `openssl s_client`.

`echo '' | openssl s_client -connect localhost:8443 -servername <my FQDN>`

```
CONNECTED(00000003)
depth=0 CN = <my FQDN>
verify error:num=20:unable to get local issuer certificate
verify return:1
depth=0 CN = <my FQDN>
verify error:num=21:unable to verify the first certificate
verify return:1
depth=0 CN = <my FQDN>
verify return:1
---
Certificate chain
 0 s:CN = <my FQDN>
   i:C = US, O = Let's Encrypt, CN = R3
---
Server certificate
-----BEGIN CERTIFICATE-----
<SNIP>
-----END CERTIFICATE-----
subject=CN = <my FQDN>

issuer=C = US, O = Let's Encrypt, CN = R3

---
No client certificate CA names sent
Peer signing digest: SHA512
Peer signature type: RSA-PSS
Server Temp Key: X25519, 253 bits
---
SSL handshake has read 1879 bytes and written 388 bytes
Verification error: unable to verify the first certificate
---
New, TLSv1.3, Cipher is TLS_AES_256_GCM_SHA384
Server public key is 2048 bit
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 21 (unable to verify the first certificate)
---
DONE
```

It seems that `CA certificate` is not send.

What I am doing wrong?

My server.toml
```
# The webserver bind address. Will use HTTPS if tls_* is provided.
# Defaults to "127.0.0.1:8443"
bindaddress = "127.0.0.1:8443"
# The read-only ldap server bind address. will use LDAPS if tls_* is provided.
# Defaults to "" (disabled)
# ldapbindaddress = "127.0.0.1:3636"
# The path to the kanidm database.
db_path = "/data/db/kanidm.db"
# If you have a known filesystem, kanidm can tune sqlite to match. Valid choices are:
# [zfs, other]
# If you are unsure about this, default to other
# zfs:
# * sets sqlite pagesize to 64k, you should set recordsize=64k on the zfs filesystem.
# db_fs_type = "zfs"
# TLS ca, certificate and key in pem format. All three must be commented, or present
tls_ca = "/data/ca.pem"
tls_cert = "/data/cert.pem"
tls_key = "/data/privkey.pem"
# The log level of the server. May be default, verbose, perfbasic, perffull
# Defaults to "default"
log_level = "default"
origin = "https://<my FQDN>"
```

### PR DESCRIPTION
**Pull request by Firstyear**
Fixes #356 - this changes from a split ca_chain/cert configuration to a single chain file. This allows rustls in tide-rustls to present the chain correctly, and allows openssl for ldaps to present the chain correctly too. it also simplifies integration to lets encrypt which provides a chain and key file by default.

- [ x ] cargo fmt has been run
- [ x ] cargo clippy has been run
- [ x ] cargo test has been run and passes
- [ x ] book chapter included (if relevant)
- [ - ] design document included (if relevant)

### CODE REVIEW

**Comment by Firstyear** [comment]
cc @vDorst

**Comment by Firstyear** [comment]
I've also pushed an updated :devel container with this change for you to test.

File: kanidm_book/src/installing_the_server.md [hunk 1]
```diff
@@ -5,41 +5,93 @@ Currently we have a pre-release docker image based on git master. They can be fo
...
+You can validate that the leaf certificate is matching to the key with the command:
+
+    # openssl rsa -noout -modulus -in key.pem | openssl sha1
+    d2188932f520e45f2e76153fbbaf13f81ea6c1ef
+    # openssl x509 -noout -modulus -in chain.pem | openssl sha1
+    d2188932f520e45f2e76153fbbaf13f81ea6c1ef
+
+You can validate the chain to the certificate with:
+
+    openssl verify -CAfile chain.pem chain.pem
```
Comment (vDorst) [inline kanidm_book/src/installing_the_server.md:57]: This seems not to work on a Let's Encrypt certificate.

Steps I use to generate chained certificate for kanidm.
```
# cd /etc/letsencrypt/live/<FQDN>
# cat cert.pem chain.pem > cert_chain.pem
# openssl verify -CAfile cert_chain.pem cert_chain.pem
C = US, O = Let's Encrypt, CN = R3
error 2 at 1 depth lookup: unable to get issuer certificate
error cert_chain.pem: verification failed
```

**Review by vDorst — CHANGES_REQUESTED** [review]
See comments in code.

**Review by Charcol — COMMENTED** [review]
Very minor suggestions, otherwise looks good

File: kanidmd/src/lib/config.rs [hunk 2]
```diff
@@ -107,21 +106,16 @@ impl Configuration {
...
+        match (chain, key) {
+            (None, None) => {}
+            (Some(chainp), Some(keyp)) => {
+                let chain = chainp.to_string();
+                let key = keyp.to_string();
+                self.tls_config = Some(TlsConfiguration { chain, key })
             }
             _ => {
-                error!("Invalid TLS configuration - must provide ca, cert and key!");
+                error!("Invalid TLS configuration - must provide chain and key!");
```
[the remark this answers is not in the document]
Comment (Firstyear) [reply kanidmd/src/lib/config.rs:118]: Good catch, changed.

File: kanidm_book/src/installing_the_server.md (hunk 1 above)
Comment (Firstyear) [reply kanidm_book/src/installing_the_server.md:60]: Good catch! I investigated why, and I have updated the documentation to suit - I have also tested on my owns lets encrypt certs and validated the steps are correct now :)

File: kanidmd/src/lib/config.rs [hunk 3]
```diff
@@ -115,7 +115,7 @@ impl Configuration {
                 self.tls_config = Some(TlsConfiguration { chain, key })
             }
             _ => {
-                error!("Invalid TLS configuration - must provide chain and key!");
+                eprintln!("Invalid TLS configuration - must provide chain and key!");
```
Comment (vDorst) [inline kanidmd/src/lib/config.rs:118]: May be add prefix `ERROR: ` so it is more inline with the other warning/error messages.

For the rest it looks good.

**Review by vDorst — APPROVED** [review]
Minor nitpick, for the rest it looks good.

File: kanidmd/src/lib/config.rs (hunk 3 above)
Comment (Firstyear) [reply kanidmd/src/lib/config.rs:118]: Good point. I'll change that and merge. Thanks for reviewing and for reporting the issue!

**Pull request merged**